wlroots: update to 0.12.0

ofiala-a51 opened the pull request:
~~There have been some important changes to wlroots that add support for the seat management daemon seatd. Since~~
1. ~~seatd still isn't available in Void repos~~
2. ~~the issue can be approached differently, which will require some discussion~~

~~this PR only updates wlroots without dealing with any of this to get the new version to end users asap. The changes required to add support for seatd shall be added in a separate PR.~~

ericonr commented:
You need to revbump / update all wlroots dependants. Afaik Wayfire isn't compatible with this version of wlroots yet.

I'm not sure we should be in a hurry, I'd like to see the wlroots 0.12.0 update already pulling in the relevant seatd stuff. We now have a good reason to merge the seatd PR.

ofiala-a51 commented:
Okay, I have added support for seatd and will look into the rest shortly. Regarding wayfire, they should support it in the next release according to [this](https://github.com/WayfireWM/wayfire/issues/858).

Regarding the changes: In addition to supporting seatd, libseat also implements majority of libelogind's functionality for use with elogind, which means libseat can be used as a backend for both seatd and elogind (afaik). This PR thus replaces wlroots' dependency libelogind with libseat since according to my testing, wlroots-based compositors work with libseat + elogind combo without issues. If there is anything I've missed, please correct me.

Note, that libseat has to be compiled with elogind support in order to work with elogind, which afaik is already done in #24580 .

ericonr commented:
You need to rebase your branch on `upstream/master`; don't use merge.

ofiala-a51 commented:
huh, I thought I rebased it... sorry, will fix

jbeich commented:
See [patch-wlroots-0.12](https://github.com/freebsd/freebsd-ports/blob/7a52419f30cea20fe8ce0c790e170e83b13b7c82/x11-wm/wayfire/files/patch-wlroots-0.12) if you'd like to update `wayfire` independently from `wlroots`.

ofiala-a51 commented:
`wayfire` already released new version that supports `wlroots-0.12.0`, so that shouldn't be a problem.

ifreund commented:
With regards to sway, emersion (the maintainer of sway/wlroots) is still deliberating  on the best course of action:
```
10:59 <emersion> (1) let distros cherry-pick 5ae4f65
10:59 <emersion> (2) cut a new minor release with only 5ae4f65
10:59 <emersion> (3) new major release, but not a fan of forcing these on each wlroots release
11:00 <FreedomBear> Of course, a new release of any kind helps us tremendously. 
11:00 <kennylevinsen> static linking would be convenient...
11:01 <emersion> with wlroots being such a intimate dependency for compositors, yeah
11:02 <columbarius> emersion:  yes. currently we upload the buffer from the gpu to ram within the compositor (screencopy) and then we copy the buffer into the pipewire buffer, depending ont the y-inverted flag flipped or not. If theres is no need to flip, we could just handover pipewire the fd of the buffer from screencopy
11:02 <FreedomBear> emersion: option 2... "With only..." You do mean all commits up to that?
11:02 <emersion> FreedomBear: no, just this commit. other commits contain new features, can't include that in a bugfix release.
11:03 <kennylevinsen> I wonder how big sway would be if we statically linked and maybe sprinkled LTO on top
11:03 <emersion> FreedomBear: basically, the bugfix release would include fixes for breaking wlroots changes, and for this particular wlroots release, we don't need any
11:03 <ammen99_> fwiw my plan is (2) if such a thing happens for wayfire
11:04 <FreedomBear> emersion: ok, so a minor release just to ensure compatibility with latest wlroots. 
11:04 <emersion> FreedomBear: exactly
11:04 <emersion> ammen99_: thx, looks like the most sensible option indeed
11:04 <emersion> kennylevinsen: wlroots is 714K it seems
11:05 <emersion> and sway 522K
11:05 <columbarius> what about a master and devel branch, cherrypicking single commits for minor compatibility releases?
11:06 <FreedomBear> emersion: any release option would be a tremendous help. Of course, a new major release is preferred, but (2) is great as well. 
11:11 <FreedomBear> Off to bed now. Thanks for entertaining our downstream requests emersion :)
11:14 <emersion> FreedomBear: np! will ping you when i've decided what to do
```

ofiala-a51 commented:
Well, until they sort it out, I made a patch. It's really just a version bump in `meson.build` since the latest version doesn't rely on any of the functionality removed in `wlroots-0.12.0`.

ofiala-a51 commented:
And with that I think this PR is ready for review. (Keeping the `WIP` since we still depend on #24580 though.)

ericonr commented on srcpkgs/wayfire/template:
```diff
@@ -1,23 +1,26 @@
+ 2b22e03d3a522baeff5798f630ffe5aa95899fd3233b291527503af5fd3e30be"
 
-build_options="elogind image"
+build_options="elogind seatd image"
```
Could this option be renamed to something like `seat_manager`?

ahesford replied:
If the option were called `setuid`, you could just drop the two `vopt_if` conditions in `depends` and use `if [ "$build_option_setuid" ]` to gate the `chmod` in `post_install`.

ericonr replied:
And then the compositors won't depend directly on `seatd` or `elogind`. I like this solution.

ofiala-a51 replied:
Should be fixed in all instances.

ericonr commented:
Please update wayfire completely: plugins-extra and wcm.

st3r4g commented:
Removing libelogind is not ok since seatd doesn't enable the elogind backend by default.
Two options:
1. build libseat with elogind support by default
2. don't remove libelogind support for wlroots
cc @ifreund

ifreund commented:
I agree that 1. makes more sense, seatd PR updated.

ofiala-a51 commented:
Completed wayfire update, updated sway, renamed suid option.

ifreund commented on srcpkgs/hikari/template:
```diff
@@ -33,5 +36,9 @@ pre_build() {
 
 post_install() {
+	if [ "$build_option_setuid" ]; then
+		# must be setuid without logind!
```
This comment doesn't make sense, seatd exists as well.

ofiala-a51 replied:
Should be fixed in all instances.

ifreund commented on srcpkgs/cage/template:
```diff
@@ -1,25 +1,25 @@
 
-build_options="elogind xwayland"
-build_options_default="xwayland"
+build_options="setuid xwayland"
```
Please add a description to `common/options.description`.

ofiala-a51 replied:
Added

ifreund reviewed (approved):
LGTM, I think this is ready to go as soon as seatd is merged.
Edit: let's reconsider this setuid build option.

ahesford commented on srcpkgs/hikari/template:
```diff
@@ -33,5 +36,9 @@ pre_build() {
+	if [ "$build_option_setuid" ]; then
+		# must be setuid for direct session management
+		chmod u+s ${DESTDIR}/usr/bin/hikari
+	fi
```
Hikari installs its executable setuid root by default, so this change is superfluous.

Regardless of whether setuid or nosetuid becomes the "standard" behavior, or people make a convincing case for preserving this build option, Hikari needs to be tested with seat management to make sure it may be launched without setuid root when seatd or elogind are used.

ofiala-a51 replied:
So there is an actual reason why the setuid bit isn't set. In that case I will undo this.

Regarding the setuid build option, I support disabling it by default as well, but @ericonr was of a different opinion, so this should probably be discussed further.

travankor replied:
I've tested hikari with elogind before and it works. The same should apply to seatd, in theory. I think the `hikari-unlocker` binary still needs to be setuid, but hikari's build system already does this.

If the setuid build option goes away, should the hikari template have something like `chmod u-s ${DESTDIR}/usr/bin/hikari` to remove setuid from the binary? Otherwise, hikari would be inconsistent with the other compositors.

ofiala-a51 replied:
That really depends on whether hikari itself (not hikari-unlocker) is able to run without suid... the project's README doesn't mention session management at all, which means someone has to actually test it

ahesford reviewed (commented):
I know I suggested renaming the `elogind` build option to `setuid` to achieve the default behavior, but looking at this propagating everywhere, I really think this is an inappropriate abuse of a build option.

After package installation, XBPS does not track file permissions, so the user can always set or clear the setuid bit depending on desired behavior. This gets a little annoying if updates toggle the bit, but is it any more annoying than having to rebuild the package after every update with some build option just to toggle the bit for you? I submit that it is *less* annoying, and users can always hold the package to avoid being surprised if they really want.

Either the binaries should default to setuid root or they should not; this is a question for regular Wayland users (one of which I am not) to hash out. But I strongly suggest the build option be dropped.

st3r4g commented:
getting this (non-fatal) error with the `elogind`:
```
00:00:00.001 [ERROR] [libseat] [libseat/backend/logind.c:605] set_type: Could not set session type: Rejected send message, 2 matched rules; type="method_call", sender=":1.6" (uid=1000 pid=534 comm="sway " label="kernel") interface="org.freedesktop.login1.Session" member="SetType" error name="(unset)" requested_reply="0" destination="org.freedesktop.login1" (uid=0 pid=481 comm="elogind-daemon " label="kernel")
```

st3r4g commented on srcpkgs/sway/template:
```diff
@@ -1,30 +1,31 @@
 	vlicense LICENSE
-	if [ -z "$build_option_elogind" ]; then
-		# must be setuid without logind!
+	if [ -z "$build_option_setuid" ]; then
```
you forgot to remove the -z here

ofiala-a51 replied:
Fixed

ifreund commented:
> getting this (non-fatal) error on `sway` with the `elogind` backend:
> 
> ```
> 00:00:00.001 [ERROR] [libseat] [libseat/backend/logind.c:605] set_type: Could not set session type: Rejected send message, 2 matched rules; type="method_call", sender=":1.6" (uid=1000 pid=534 comm="sway " label="kernel") interface="org.freedesktop.login1.Session" member="SetType" error name="(unset)" requested_reply="0" destination="org.freedesktop.login1" (uid=0 pid=481 comm="elogind-daemon " label="kernel")
> ```
> 
> the compositor starts anyway and seems to work normally

This is due to elogind not supporting the `SetType` method yet. It will be supported in `v246`, added in [this commit](https://github.com/elogind/elogind/commit/490783b27f38c5e96523ea3aa130a78bb38b5814)

ifreund commented:
I agree with @ahesford's assessment of the `setuid` situation; the build option should go. I would be in favor of *not* producing setuid binaries as I see using a seat management daemon such as `elogind` or `seatd` as the reasonable thing for most users to do.

sgn commented:
> I know I suggested renaming the `elogind` build option to `setuid` to achieve the default behavior, but looking at this propagating everywhere, I really think this is an inappropriate abuse of a build option.

`xorg-server` also has `elogind` build option, with that option on, `setsuid` is dropped, too. So, I think *the* convention is use `elogind`.

> 
> After package installation, XBPS does not track file permissions, so the user can always set or clear the setuid bit depending on desired behavior. This gets a little annoying if updates toggle the bit, but is it any more annoying than having to rebuild the package after every update with some build option just to toggle the bit for you? I submit that it is _less_ annoying, and users can always hold the package to avoid being surprised if they really want.
> 
> Either the binaries should default to setuid root or they should not; this is a question for regular Wayland users (one of which I am not) to hash out. But I strongly suggest the build option be dropped.

ericonr commented:
If we were to switch to no-suid by default, `cage`, `wayfire` and `sway` should get an `INSTALL.msg` about it (it's a good opportunity to clean wayfire's current message as well).

ofiala-a51 commented:
So I take it that everyone supports removal of the setuid build option, nice.

I have tried running latest hikari as a non-privileged user with seatd and without suid and it seems to work without any issues. So @ahesford would it be okay if I add the `WITHOUT_SUID=YES` build param to disable it?

ahesford commented:
On 2020-11-11 at 09:03 (UTC -0500), ofiala-a51 wrote:
> So I take it that everyone supports removal of the setuid build option, nice.
> 
> I have tried running latest hikari as a non-privileged user with seatd and without suid and it seems to work without any issues. So @ahesford would it be okay if I add the `WITHOUT_SUID=YES` build param to disable it?

That's fine with me.

ofiala-a51 commented:
Removed SUID, added install messages notifying users about this change. Also removed old install message for wayfire as it was basically just short documentation, which is a use of install messages that has been retired.

ericonr commented:
Please patch 'wofi' as well https://todo.sr.ht/~scoopta/wofi/137

ofiala-a51 commented:
Fixed.

ericonr commented:
You're missing the `wf-shell` update. Remember to include the `wayland-logout` distfile.

ericonr commented:
Looks a little something like this:

```
diff --git a/srcpkgs/wf-shell/template b/srcpkgs/wf-shell/template
index f88d967d17..a9db707ad1 100644
--- a/srcpkgs/wf-shell/template
+++ b/srcpkgs/wf-shell/template
@@ -1,9 +1,12 @@
 # Template file for 'wf-shell'
 pkgname=wf-shell
-version=0.5.0
+version=0.6.1
 revision=1
+# not using the exact upstream commit because this one includes the license
+_logout_hash=c9f0baea901e50d3b5c2e0e0a53958cb2b930f14
 build_style=meson
 build_helper="gir"
+configure_args="-Dwayland-logout:implementation=c"
 hostmakedepends="gobject-introspection pkg-config wayland-devel"
 makedepends="alsa-lib-devel pulseaudio-devel gtkmm-devel wayfire-devel
  gtk+3-devel gtk-layer-shell-devel"
@@ -11,10 +14,18 @@ short_desc="Wayfire shell with GTK-based panel and background client"
 maintainer="Young Jin Park <youngjinpark20@gmail.com>"
 license="MIT"
 homepage="https://wayfire.org"
-distfiles="https://github.com/WayfireWM/wf-shell/releases/download/${version}/wf-shell-${version}.tar.xz"
-checksum=fa481e867784b94b08857ec7bd27bf43251490d5d50c1e834e7ea64d908792f0
+distfiles="https://github.com/WayfireWM/wf-shell/releases/download/v${version}/wf-shell-${version}.tar.xz
+ https://github.com/soreau/wayland-logout/archive/${_logout_hash}.tar.gz"
+checksum="da8e5a16dde8f44966d41449452bcc7bc2466d5945324aca714ecc82827c884a
+ 455564d8bf7c37d0f77a55402d9d96173419bd08617bc4d0bef313a62ab82d49"
+
+post_extract() {
+	rmdir ${wrksrc}/subprojects/wayland-logout
+	mv wayland-logout-${_logout_hash} ${wrksrc}/subprojects/wayland-logout
+}
 
 post_install() {
 	vlicense LICENSE
+	vlicense subprojects/wayland-logout/LICENSE
 	vsconf wf-shell.ini.example
 }
```

I'm still testing it out, tho.

ericonr commented:
Wayfire is running fine with this PR, besides the possible reapperance of a bug I thought had been solved.

ericonr commented:
@ofiala-a51  Please pull in my suggestions and let CI run; seatd has been merged.

ericonr commented:
Squash the two wayfire commits and force push into the PR, please.
